src/expr/impl/src/scalar/case.rs

```diff
@@ -106,6 +107,135 @@ impl Expression for CaseExpression {
     }
 }
 
+/// With large scale of simple form match arms in case-when expression,
+/// we could optimize the `CaseExpression` to `ConstantLookupExpression`,
+/// which could significantly facilitate the evaluation of case-when.
+#[derive(Debug)]
+struct ConstantLookupExpression {
+    return_type: DataType,
+    arms: HashMap<ScalarImpl, BoxedExpression>,
+    fallback: Option<BoxedExpression>,
+    /// `operand` must exist at present
+    operand: BoxedExpression,
+}
+
+impl ConstantLookupExpression {
+    fn new(
+        return_type: DataType,
+        arms: HashMap<ScalarImpl, BoxedExpression>,
+        fallback: Option<BoxedExpression>,
+        operand: BoxedExpression,
+    ) -> Self {
+        Self {
+            return_type,
+            arms,
+            fallback,
+            operand,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Expression for ConstantLookupExpression {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let input_len = input.capacity();
+        let mut builder = self.return_type().create_array_builder(input_len);
+
+        // Evaluate the input DataChunk at first
+        let eval_result = self.operand.eval(input).await?;
+
+        for i in 0..input_len {
+            let datum = eval_result.datum_at(i);
+            let (row, vis) = input.row_at(i);
+
+            // Check for visibility
+            if !vis {
+                builder.append_null();
+                continue;
+            }
+
+            // Note that the `owned_row` here is extracted from input
+            // rather than from `eval_result`
+            let owned_row = row.into_owned_row();
+
+            if let Some(expr) = self.arms.get(datum.as_ref().unwrap()) {
+                builder.append(expr.eval_row(&owned_row).await.unwrap().as_ref());
+            } else {
+                // Otherwise this should goes to the fallback arm
+                // The fallback arm should also be const
+                if let Some(ref fallback) = self.fallback {
+                    builder.append(fallback.eval_row(&owned_row).await.unwrap().as_ref());
+                } else {
+                    builder.append_null();
+                }
+            }
+        }
+
+        Ok(Arc::new(builder.finish()))
+    }
+
+    async fn eval_row(&self, input: &OwnedRow) -> Result<Datum> {
```

Comment:
This method should also use operand to evaluate and doesn't need to care about the constant folding input here.

Reply:
Oops, forget about this one. 😄

Reply:
fixed!